### ISSUE TITLE
Lerna version canary support

### ISSUE THREAD

**Issue by JonnyBGod** [lerna/lerna#1583]
Since version 3.0.0 there is no way to bump package version with canary hash.

This is very helpful in some use cases such as nightly builds where we want to skip git push and tags but still need to distinguish builds.

Without it we will end up with multiple ```0.0.1.develop.0``` builds.

**Comment by evocateur** [comment]
I'm not sure why you need to run the same nightly build repeatedly? The SHA is appended with a + to the version when publishing a canary release, which _does_ require un-shallow git tags to properly increment the version.

**Comment by evocateur** [comment]
That sounds very complicated. Lerna manages npm packages, not Docker images.

**Comment by JonnyBGod** [comment]
I understand that, but with a bit of extra work, up until now lerna was flexible enough to use for npm packages as well as other types of packages while keeping all versioning concise.

Now what I do not understand is why do you support canary releases with the publish command but not with version command. Why not move that functionality entirely to the version command.

The problem come from the combination of deprecating ```--skip-npm``` and  not supporting ```lerna version --canary```.

Would you consider re-enabling it?

**Comment by evocateur** [comment]
I am receptive to the argument for extracting the canary version generation logic into a separate package (I'll probably end up doing it anyway to support a planned `lerna publish release` subcommand), but I do not see `lerna version --canary` happening.

TL;DR: In lerna v3, `lerna publish` makes changes to npm registries and `lerna version` makes changes to git repositories. `--canary` never makes changes to git repositories, therefore it belongs in the command that makes changes to npm registries.

Lerna v2 conflated the two, (not coincidentally) like `yarn publish` does with `npm version` and `npm publish`. One problem with v2's approach was exemplified by the `--skip-npm` and `--skip-git` options themselves, contributing to a gargantuan morass of increasingly unmaintainable conditional statements. I took the opportunity of a major version to cut that Gordian knot.

The v2 `--canary` option was another instance of an overloaded mess of conflicting semantics. It had its own method of setting a `--preid`, but no possible way to safely use `^` ranges because it didn't produce semver prerelease versions compatible with npm's resolution algorithm. It was poorly tested, difficult to support, and not especially useful, at the end of the day. Separating concerns into two commands helped clarify what `--canary` was _for_.

I made a great effort in v3 to provide composable modules (almost 50 of them, so far) for folks to reuse and remix various bits of Lerna logic for their specific, perhaps esoteric, use cases. It's not perfect, but suggestions/PRs for module extraction are always welcome. It is certainly one of my goals to facilitate workflows that I don't believe belong in core via useful and expressive utility packages. The documentation _definitely_ needs a lot of help, and I apologize for that.

**Comment by JonnyBGod** [comment]
Thank for the explanation. In general terms I totally agree with you.

But then again, it looks to me that there should be a way to calculate the new version for different use cases without the need to actually publish anything.

This seams like a reasonable use case, to want to mix different packaging needs while keeping all versions concise.
For that I would rather make sure that everything builds up before actually starting to publish the new packages.

I can see several different forms this could take form. Maybe a some kind of ```--dry-run``` flag would be appropriated. Or detaching the version calculation completely. Or enabling ```lerna changed``` to produce results with the calculated next version.

Do any of these sounds like doable  to you?

**Comment by evocateur** [comment]
`--dry-run`: #51

Ensuring everything builds and tests pass is as simple as running `<my-build-script> && npm t` before your `lerna publish` execution. Or even running it during the root `prepublishOnly` lifecycle, for example.

Canary version calculation is [not especially complicated](https://github.com/lerna/lerna/blob/2760306a68183fcaffa4a0e4d786ce4e46956f62/commands/publish/index.js#L211-L218), working from [parsed `git describe` output](https://github.com/lerna/lerna/blob/2760306a68183fcaffa4a0e4d786ce4e46956f62/utils/describe-ref/lib/describe-ref.js#L53-L63). The "not especially complicated" bits do need extraction, like I said above.

`lerna changed` could gain that ability, I suppose. As it is right now, it's basically a shortcut for `lerna ls --since`, in terms of its actual output. (oh, and it exits non-zero when there are no changes, so that's different...)

**Comment by sbaechler** [comment]
I just ran into a problem with our private Nexus npm server. The current version does not support npm tags ([issue](https://issues.sonatype.org/browse/NEXUS-9862)). Therefore it is not possible to reference the latest build using "latest".

It would really help if there was an alternative canary versioning using build count instead of hash (e.g. using `git rev-list HEAD --count`). This would make it possible to have an incrementing version without having to create a commit.

**Comment by evocateur** [comment]
@sbaechler Canary version computation does not use npm [dist-tags](https://docs.npmjs.com/cli/dist-tag) (like `latest`), it uses git tags (via `git describe`, as I linked to earlier) to determine the number of commits since the last release.

**Comment by sbaechler** [comment]
@evocateur Sure, but as I understand it, npm tags are currently the only way to get the latest version of a canary build (in another project). The hashes are not sequential so the usual `^` versioning does not work.

**Comment by evocateur** [comment]
> The hashes are not sequential so the usual `^` versioning does not work.

@sbaechler This is not the case since Lerna v3.0.0 (see #1522 for details).

As #1614 highlighted, there is still and edge case when you are attempting to publish a canary release using the default "fixed" versioning scheme _before_ you've made _any_ formal tagged releases. I fixed the independent fallback, but it's a static `1` default instead of a perhaps more appropriate `git rev-list HEAD --count`.

So right now, as of the Lerna v3, `lerna publish --canary` with existing tagged releases will always yield properly-incrementing prerelease version ranges.

**Comment by sbaechler** [comment]
Ok, I guess this is the better approach. I was trying to avoid a git push from CI back to the repo because setting this up in Gitlab CI is not that smooth yet. But Gitlab adding version Tags would actually make for a better history.

Is there a best practice example for a build pipeline with a Lerna project? Otherwise I would create one.

**Comment by pksunkara** [comment]
I think a better way of rephrasing the whole issue would be the following:

[babel/minify](https://github.com/babel/minify) used lerna@2 to publish canary on CI (for every commit) until recently. I had made a PR to upgrade lerna to v3 (https://github.com/babel/minify/pull/900 & https://github.com/babel/minify/pull/901). As you can see from the CI log [here](https://travis-ci.com/babel/minify/jobs/147386203), it successfully published `0.5.0-alpha.9+7002b8c`. Looks good.

But now, they added more commits to the master branch and the CI ran again [here](https://travis-ci.com/babel/minify/jobs/147485107). As you can see, lerna canary publish decided that the tag would be `0.5.0-alpha.9+f00c588`. Only the SHA changed, not the `alpha.9` part.

This is because of travis-ci doing a shallow clone of the repo. The only way to fix this would be doing, `git fetch --unshallow` before the lerna publish step. @evocateur You should probably document this fix somewhere.

**Comment by TomLingham** [comment]
Is there any intention to support `lerna version --canary`?
For our use case, we want to be able to canary version and pack all our packages into tars so that we can publish them at a later time. Especially useful for local verification where the actual publish step resides in CI.
